File: BookStore_Frontend/src/services/address.service.ts
import apiClient from './api.client';
import {
  type AddressRequest,
  type AddressResponse,
} from '../types';

class AddressService {
  private readonly BASE_URL = '/addresses';

  async getUserAddresses(): Promise<AddressResponse[]> {
    const response = await apiClient.get<AddressResponse[]>(this.BASE_URL);
    return response.result ?? [];
  }

  async createAddress(data: AddressRequest): Promise<AddressResponse> {
    const response = await apiClient.post<AddressResponse>(this.BASE_URL, data);
    if (!response.result) {
      throw new Error(response.message || 'Failed to create address');
    }
    return response.result;
  }

  async updateAddress(addressId: string, data: AddressRequest): Promise<AddressResponse> {
    const response = await apiClient.put<AddressResponse>(
      `${this.BASE_URL}/${addressId}`,
      data
    );
    if (!response.result) {
      throw new Error(response.message || 'Failed to update address');
    }
    return response.result;
  }

  async deleteAddress(addressId: string): Promise<void> {
    await apiClient.delete<unknown>(`${this.BASE_URL}/${addressId}`);
  }

  async setDefaultAddress(addressId: string): Promise<AddressResponse> {
    const response = await apiClient.patch<AddressResponse>(
      `${this.BASE_URL}/${addressId}/default`
    );
    if (!response.result) {
      throw new Error(response.message || 'Failed to set default address');
    }
    return response.result;
  }

  async getDefaultAddress(): Promise<AddressResponse | null> {
    const response = await apiClient.get<AddressResponse>(
      `${this.BASE_URL}/default`
    );
    return response.result ?? null;
  }

  async countUserAddresses(): Promise<number> {
    const response = await apiClient.get<number>(`${this.BASE_URL}/count`);
    return response.result ?? 0;
  }

  async deleteAllUserAddresses(): Promise<void> {
    await apiClient.delete<unknown>(`${this.BASE_URL}/all`);
  }
}

export const addressService = new AddressService();
export default addressService;

File: BookStore_Backend/src/main/java/chucnguyen/bookstore/service/AddressService.java
package chucnguyen.bookstore.service;

import chucnguyen.bookstore.dto.request.AddressRequest;
import chucnguyen.bookstore.dto.response.AddressResponse;
import chucnguyen.bookstore.entity.Address;
import chucnguyen.bookstore.entity.User;
import chucnguyen.bookstore.exception.AppException;
import chucnguyen.bookstore.exception.ErrorCode;
import chucnguyen.bookstore.mapper.AddressMapper;
import chucnguyen.bookstore.repository.AddressRepository;
import chucnguyen.bookstore.repository.UserRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class AddressService {

    private final AddressRepository addressRepository;
    private final UserRepository userRepository;
    private final AddressMapper addressMapper;

    @Cacheable(value = "addresses", key = "'user_' + #email")
    public List<AddressResponse> getUserAddresses(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        List<Address> addresses = addressRepository.findByUserIdOrderByIsDefaultDescCreatedAtDesc(user.getId());
        return addressMapper.toAddressResponseList(addresses);
    }

    @Transactional
    @CacheEvict(value = "addresses", allEntries = true)
    public AddressResponse createAddress(String email, AddressRequest request) {
        log.info("Creating new address for user: {}", email);

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        Address address = addressMapper.toAddress(request);
        address.setUser(user);

        // Save first to get ID
        address = addressRepository.save(address);

        // Then unset others if this is default
        if (Boolean.TRUE.equals(request.getIsDefault())) {
            addressRepository.unsetOtherDefaults(user.getId(), address.getId());
        }

        log.info("Address created successfully for user: {}", email);

        return addressMapper.toAddressResponse(address);
    }

    @Transactional
    @CacheEvict(value = "addresses", allEntries = true)
    public AddressResponse updateAddress(String email, String addressId, AddressRequest request) {
        log.info("Updating address {} for user: {}", addressId, email);

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new AppException(ErrorCode.ADDRESS_NOT_FOUND));

        // Check ownership
        if (!address.getUser().getId().equals(user.getId())) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        // Update
        addressMapper.updateAddressFromRequest(request, address);

        // Handle default flag
        if (Boolean.TRUE.equals(request.getIsDefault())) {
            addressRepository.unsetOtherDefaults(user.getId(), addressId);
        }

        address = addressRepository.save(address);

        return addressMapper.toAddressResponse(address);
    }

    @Transactional
    @CacheEvict(value = "addresses", allEntries = true)
    public void deleteAddress(String email, String addressId) {
        log.info("Deleting address {} for user: {}", addressId, email);

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new AppException(ErrorCode.ADDRESS_NOT_FOUND));

        // Check ownership
        if (!address.getUser().getId().equals(user.getId())) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        addressRepository.delete(address);

        log.info("Address deleted successfully");
    }

    @Transactional
    @CacheEvict(value = "addresses", allEntries = true)
    public AddressResponse setDefaultAddress(String email, String addressId) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new AppException(ErrorCode.ADDRESS_NOT_FOUND));

        // Check ownership
        if (!address.getUser().getId().equals(user.getId())) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        // Unset other defaults
        addressRepository.unsetOtherDefaults(user.getId(), addressId);

        // Set as default
        address.setIsDefault(true);
        address = addressRepository.save(address);

        return addressMapper.toAddressResponse(address);
    }

    @Cacheable(value = "addresses", key = "'default_' + #email")
    public AddressResponse getDefaultAddress(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        Address address = addressRepository.findByUserIdAndIsDefaultTrue(user.getId())
                .orElseThrow(() -> new AppException(ErrorCode.ADDRESS_NOT_FOUND,
                        "No default address found"));

        return addressMapper.toAddressResponse(address);
    }

    @Cacheable(value = "addresses", key = "'count_' + #email")
    public long countUserAddresses(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        return addressRepository.countByUserId(user.getId());
    }

    @Transactional
    @CacheEvict(value = "addresses", allEntries = true)
    public void deleteAllUserAddresses(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        addressRepository.deleteByUserId(user.getId());
    }
}
